Extend yum module to support downgrading packages

willusher opened the pull request:
This re-introduces yum package downgrades that was removed in 1.9.1
(commit d77f197) due to a bug.

Resolves issue #1419

Commit:
Extend yum module to support downgrading packages

This re-introduces yum package downgrades that was removed in 1.9.1
(commit d77f197) due to a bug.

Resolves issue #1419

robynbergeron commented:
Thanks @willusher for this PR. This module is maintained by the Ansible core team, so it can take a while for patches to be reviewed. Thanks for your patience.

copumpkin commented:
:+1: :+1: on the functionality, although I haven't yet tried @willusher's implementation. Is it just an adaptation of the original reverted commit? What was the bug that caused it to fail the first time around?

willusher commented:
@copumpkin - The original implementation was done before a major rewrite of the yum module, so this is a fairly heavy adaptation. The bug in the original code was the `splitFilename()` function returning unexpected results in certain scenarios.

ryansydnor commented:
I just spent a little while trying to figure out when a downgrade is triggered. Looks like it happens if/when you specify a file version. Can we add that to the documentation?

willusher commented:
Thanks @ryansydnor - Added line in [documentation](https://github.com/ansible/ansible-modules-core/commit/99cbccfe70d53783895ee53b2bc2a04644f9821e#diff-50dbb2224ca8ab1fc7b6b693f9be32faR50)

anonymuse commented:
:+1: here too, this would be helpful to have again.

ryansydnor commented:
Thanks @willusher for the docs! :shipit:

joelthompson commented:
:+1: I think this would be really good functionality to have back.

gregdek commented:
Thanks @willusher for this PR. Unfortunately, it is not mergeable in its current state due to merge conflicts. Please rebase your PR. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

gregdek commented:
Thanks @willusher for this PR. This module is maintained by the Ansible core team, so it can take a while for patches to be reviewed. Thanks for your patience.

aelsabbahy commented:
When there are no changes, I get the following exception:

```
Traceback (most recent call last):                                                                                     
  File "/home/ec2-user/.ansible/tmp/ansible-tmp-1455645028.98-220775352304363/yum", line 3454, in <module>             
    main()                                                                                                             
  File "/home/ec2-user/.ansible/tmp/ansible-tmp-1455645028.98-220775352304363/yum", line 1062, in main                 
    disablerepo, disable_gpg_check, exclude, repoquery)                                                                
  File "/home/ec2-user/.ansible/tmp/ansible-tmp-1455645028.98-220775352304363/yum", line 978, in ensure                
    res = install(module, pkgs, repoq, yum_basecmd, conf_file, en_repos, dis_repos)                                    
  File "/home/ec2-user/.ansible/tmp/ansible-tmp-1455645028.98-220775352304363/yum", line 711, in install               
    res = exec_install(module, items, 'install', pkgs, res, yum_basecmd, tempdir)                                      
  File "/home/ec2-user/.ansible/tmp/ansible-tmp-1455645028.98-220775352304363/yum", line 564, in exec_install          
    out = '%s: Nothing to do' % spec                                                                                   
UnboundLocalError: local variable 'spec' referenced before assignment
```

[culprit](https://github.com/ansible/ansible-modules-core/pull/2744/files#diff-50dbb2224ca8ab1fc7b6b693f9be32faR574)

willusher commented:
Thanks @aelsabbahy - It looks like that code path is only hit when specifying a package group (e.g: @Developer Tools) that is already installed. Anyway, the status of each package within the group is already printed individually, so it didn't make sense to print the list again, so I just removed that.

gregdek commented:
Thanks @willusher for this PR. Unfortunately, it is not mergeable in its current state due to merge conflicts. Please rebase your PR. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

gregdek commented:
Thanks @willusher for this PR. This module is maintained by the Ansible core team, so it can take a while for patches to be reviewed. Thanks for your patience.

Core team: please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with 'needs_revision' or merge as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

abadger commented:
Discussed at today's meeting.  Came up with some corner cases in what we could pass in as a package spec that could cause issues for splitFilename(): https://gist.github.com/abadger/c673b5c02f15c775603ddeca8bbdd348

abadger commented:
The question was asked on irc if we could implement the regex from distutils in order to parse the rpm NEVRAs... I don't think that's going to work  python version strings are very simple compared to rpm NEVRAs.

I think maybe what we need is to be able to either filter our input to the splitFilename() (detecting cases that are odd before we pass it to it) or be able to identify our failure cases afterwards better.

willusher commented:
@abadger - I looked through the code again, and realized that I didn't need to parse the version number from user input. I could let yum do that and just extract the NEVRA metadata from pkglist.

Of course I figured that out _after_ writing a regex to parse the package name and version number, so I'll just leave this here in case anyone happens to need that:
`r'(^[\w-]+[^\d])-(\d[\w.:-]+)$'`

abadger commented on packaging/os/yum.py:
```diff
@@ -646,60 +698,40 @@ def install(module, items, repoq, yum_basecmd, conf_file, en_repos, dis_repos):
+            # determine if this is a downgrade, and then use the yum downgrade command to install the RPM.
+
+            # Get the NEVRA of the requested package
+            (name, ver, rel, epoch, arch) = splitFilename(pkglist.pop())
```
Why use pkglist.pop() insgtead of pkglist[0]?

is pkglist[0] always going to be the name that was specified?  What if the name that was specified is a virtual provide?  ie: Package named foo at version 2.0 has Provides: bar = 1.5.  If the ansible command specifies name: "bar=1.0" will pkglist[0] be foo-2.0 or bar-1.5?

abadger commented on packaging/os/yum.py:
```diff
@@ -646,60 +698,40 @@ def install(module, items, repoq, yum_basecmd, conf_file, en_repos, dis_repos):
+
+            # Check if any version of the requested package is installed
+            inst_pkgs = is_installed(module, repoq, name, conf_file, en_repos=en_repos, dis_repos=dis_repos, is_pkg=True)
+            if inst_pkgs:
```
It seems like this code may never be hit.  We already called is_installed() before.  If that showed as installed we go to the next iteration of the loop.  This looks like it may be checking the same name with is_installed and therefore it can only be False at this point.  What am I missing?

abadger commented on packaging/os/yum.py:
```diff
@@ -646,60 +698,40 @@ def install(module, items, repoq, yum_basecmd, conf_file, en_repos, dis_repos):
-        # Record change
-        res['changed'] = changed
+        res = exec_install(module, items, 'install', pkgs, res, yum_basecmd, tempdir)
 
```
Looking at this I realized that we're breaking the install ansible command into two yum actions.  This is counter to what we did in 48422fba85b73f39ad48b26e7e104b0e84064cf2

I think we don't encounter the most common case that that commit addressed -- when a package isn't installed already and you ask to install a different one that depends upon it -- but there are other thing that are... changed by doing this two-step:

(1) we can end up fulfilling half of what was requested if the downgrade can succeed but the subsequent install fails
(2) we can end up with the second half of the ansible request undoing what the first part accomplished -- for instance, if the first part says foo=1.0 and the second part says install bar where bar has a dependency on foo-2.0.    Each run of the playbook will end up with ansible downgrading foo to 1.0 and then upgrading it to 2.0.  No error will be reported.

jchesshir replied:
It seems to me that if a version is named explicitly, we should accept that the user wants that version even if it does conflict with the dependencies of another.  If possible, there should be a warning about the conflict, but the true failure should come when the applications try to run if the user didn't know what they were doing.  

Keying off abadger's example, if the user stated that foo_1.0 was to be installed despite bar depending on foo_2.0, we should assume that the user knows why bar is depending on foo_2.0 and has determined that not to be a problem for their particular use case.  We can only hand hold an incompetent user for so long.  If they decide to override our cautions, they will have to fall on their face if they haven't done their homework.

Not sure how easy that is to implement, but that's my two cents.

willusher commented on packaging/os/yum.py:
```diff
@@ -646,60 +698,40 @@ def install(module, items, repoq, yum_basecmd, conf_file, en_repos, dis_repos):
-        # Record change
-        res['changed'] = changed
+        res = exec_install(module, items, 'install', pkgs, res, yum_basecmd, tempdir)
 
```
As far as I can tell, there's no way to include the downgrades and upgrades in the same yum transaction. Maybe test if the end state matches the spec and error out of it doesn't? That still doesn't really solve the problem.

gregdek commented:
Thanks @willusher for this PR. Unfortunately, it is not mergeable in its current state due to merge conflicts. Please rebase your PR. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

gregdek commented:
@willusher A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

gregdek commented:
@willusher Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

gregdek commented:
@willusher A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

willusher commented:
ready_for_review

gregdek commented:
Thanks @willusher for this PR. This module is maintained by the Ansible core team, so it can take a while for patches to be reviewed. Thanks for your patience.

Core team: please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with 'needs_revision' or merge as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

Commit:
Make Ansible's IAM module properly delete roles

Previously this module was not reliably deleting IAM roles, and would
return 'BotoServerError: 409 Conflict'. Now it follows the procedure
outlined in AWS official docs to delete a role.

See: http://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_manage_delete.html#roles-managingrole-deleting-api

Adding waits does change the behavior of this module, but in our testing
the role deletion did not take any additional time. The waits will improve
the cases where we are experiencing AWS issues, which have caused
problems in the past.

gregdek commented:
Thanks @willusher for this PR. Unfortunately, it is not mergeable in its current state due to merge conflicts. Please rebase your PR. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

For help on how to do this cleanly please see http://docs.ansible.com/ansible/community.html#contributing-code-features-or-bugfixes

[This message brought to you by your friendly Ansibull-bot.]

joelthompson commented:
Hi @gregdek -- I apologize if I come across as frustrated, but this is now the fourth time that the PR has not been mergeable due to upstream changes. It seems like a waste of time to repeatedly ask @willusher to update this PR only to have it get ignored and languish. This just doesn't seem to be a fair way to treat the community members who are trying to make Ansible better.

Can you just be upfront about whether this will get reviewed if it gets updated again?

gregdek commented:
Some PRs are more challenging than others, and the fact that this one hits both iam and yum makes it more challenging than most. 

My advice to @willusher is to attend an upcoming Core Packaging meeting (https://github.com/ansible/community/blob/master/MEETINGS.md) -- it exists as a venue to tackle precisely this kind of complex issue.

gregdek commented:
@willusher A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

jchesshir commented:
Any word on when this can be released.  My team is also really wanting this functionality.

gregdek commented:
@willusher Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

gregdek commented:
@willusher A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

gregdek commented:
@willusher Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

ghost commented:
This is very basic functionality that is absolutely needed in Ansible. My suggestion is for the core team to tackle this problem and be done with it. It has almost been a year now. Right now the Yum module is not usable, and I'd even call it broken.

jchesshir commented:
In an effort to speed things along, I have cloned this repo, pulled in Will Usher's changes, and rebased them over the latest version of the devel branch.  I then created another pull request and copied it over to the new repository.  You can find the migrated pull request here: https://github.com/ansible/ansible/pull/19567.

@willusher,
If you can at all find the time to attend one of the meetings as suggested by @gregdek back in September, please do so ASAP.  https://github.com/ansible/community/blob/master/MEETINGS.md.  There are a lot of us out here who want to see this resolved.

I used my best judgement to try to work your changes into their code where it made sense, but I am far from being expert enough to be able to even compile or run it, much less defend the changes to the administrators.

Thank you
